cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/ResourceDistributionGoal.java

```diff
@@ -462,11 +727,139 @@ private boolean isLoadUnderBalanceUpperLimitAfterChange(ClusterModel clusterMode
     }
   }
 
+  /**
+   * Check if moving the source replica to destination broker would decrease the utilization difference between source
+   * and destination brokers.
+   *
+   * @param sourceReplica Source replica to be moved.
+   * @param destinationBroker Destination broker to receive the source replica utilization.
+   * @return True if the change would lead to a better balance, false otherwise.
+   */
+  private boolean isAcceptableAfterReplicaMove(Replica sourceReplica, Broker destinationBroker) {
+    double sourceUtilizationDelta = 0 - sourceReplica.load().expectedUtilizationFor(resource());
+
+    double destinationBrokerUtilization = destinationBroker.load().expectedUtilizationFor(resource());
+    return isGettingMoreBalanced(sourceReplica, sourceUtilizationDelta, destinationBrokerUtilization);
+  }
+
+  /**
+   * Check if the swap would decrease the utilization difference between source and destination brokers.
+   *
+   * @param sourceReplica Source replica to be swapped.
+   * @param destinationReplica Destination replica to be swapped.
+   * @return True if the change would lead to a better balance, false otherwise.
+   */
+  private boolean isSelfSatisfiedAfterSwap(Replica sourceReplica, Replica destinationReplica) {
+    double sourceUtilizationDelta = destinationReplica.load().expectedUtilizationFor(resource())
+                                    - sourceReplica.load().expectedUtilizationFor(resource());
+
+    double destinationBrokerUtilization = destinationReplica.broker().load().expectedUtilizationFor(resource());
+    return isGettingMoreBalanced(sourceReplica, sourceUtilizationDelta, destinationBrokerUtilization);
+  }
+
+  /**
+   * Check if the utilization difference between source and destination brokers would decrease after the source
+   * utilization delta is removed from the destination and added to source broker.
+   *
+   * @param sourceReplica Source replica.
+   * @param sourceUtilizationDelta Utilization that would be removed from the destination and added to source broker.
+   * @param destinationBrokerUtilization The utilization of destination broker.
+   * @return True if the change would lead to a better balance, false otherwise.
+   */
+  private boolean isGettingMoreBalanced(Replica sourceReplica, double sourceUtilizationDelta, double destinationBrokerUtilization) {
+    double sourceBrokerUtilization = sourceReplica.broker().load().expectedUtilizationFor(resource());
+
+    double prevDiff = sourceBrokerUtilization - destinationBrokerUtilization;
+    double nextDiff = prevDiff + (2 * sourceUtilizationDelta);
+
+    return Math.abs(nextDiff) < Math.abs(prevDiff);
+  }
+
+  private boolean isLoadStillWithinLimitsAfterSwap(Replica sourceReplica, Replica destinationReplica) {
```

Comment:
This can likely be merged with `isHostLoadStillWithinLimitsAfterSwap ` by using lambda.